### PR TITLE
Unroll Untraced Execution

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Add a second run-to-halt path used when there is no tracing.  This path calls static final versions of select operations with the aim of avoiding the megamorphic dispatch of the original loop.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Comment by shemnon** [comment]
~20-30% from prior PR, 
~41-48% from 22.7.x

The hitch is that these are only for executions that do not involve tracing.  This is standard for a mainnet production, but hedera depends on a tracer for some execution.  I think I can re-work the trace interface to do what hedera needs.

Old (#4533)
```
{"output":"e2ef8b895ee03259ee57850b1a1ed4aa7689bec352659200e63440303f9d0b81","Mgps":"211.153","gasUsed":"0x96cd6dc2","time":11982032500,"test":"loopMul","fork":"Berlin","d":0,"g":0,"v":0,"postHash":"0x65134195701ec3aa354731a26d9533e946f11e11e60247cc96f57b305778fb12","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"ffffffffffffffffffffffffffaaffffffffffffffffbbffffffffffffff0009","Mgps":"235.737","gasUsed":"0x1778312c4","time":26724899042,"test":"loopMul","fork":"Berlin","d":1,"g":0,"v":0,"postHash":"0xa419240a3f3b2a7858212be7e9d6418d8cffd1673dca395582f191621555993d","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"0e1c6aac6663c379a52d9ccc7ba4757131020772d41447dfcf478cf9fb0c2bbf","Mgps":"147.324","gasUsed":"0x212496c6","time":3774310042,"test":"loopMul","fork":"Berlin","d":2,"g":0,"v":0,"postHash":"0x0ed14414e692f2ed97cef2b5a500863e2de358b42351d93b963ef4efb3d13d72","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"e2ef8b895ee03259ee57850b1a1ed4aa7689bec352659200e63440303f9d0b81","Mgps":"212.748","gasUsed":"0x96cd658e","time":11892231375,"test":"loopMul","fork":"Istanbul","d":0,"g":0,"v":0,"postHash":"0xee92d0c049225c53aa9b156986ad4e7cc9ec2524395c60876a17259fadc8aeae","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"ffffffffffffffffffffffffffaaffffffffffffffffbbffffffffffffff0009","Mgps":"235.210","gasUsed":"0x177830a90","time":26784780875,"test":"loopMul","fork":"Istanbul","d":1,"g":0,"v":0,"postHash":"0xe9bca424874ebc2457c9822f6c2f6d2b81ca9194b767c62f9b3c1e554d0a58f3","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"0e1c6aac6663c379a52d9ccc7ba4757131020772d41447dfcf478cf9fb0c2bbf","Mgps":"150.844","gasUsed":"0x21248e92","time":3686206708,"test":"loopMul","fork":"Istanbul","d":2,"g":0,"v":0,"postHash":"0xdf31131823f2d621cf2c2d980084681c7ac75a3991480c7d05a7031df34e21c5","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"e2ef8b895ee03259ee57850b1a1ed4aa7689bec352659200e63440303f9d0b81","Mgps":"214.435","gasUsed":"0x96cd6dc2","time":11798662084,"test":"loopMul","fork":"London","d":0,"g":0,"v":0,"postHash":"0x662588feffcd4391bc440510de21cbfa8f6c59ceafb2384ed706eccb55ec3a92","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"ffffffffffffffffffffffffffaaffffffffffffffffbbffffffffffffff0009","Mgps":"237.000","gasUsed":"0x1778312c4","time":26582431333,"test":"loopMul","fork":"London","d":1,"g":0,"v":0,"postHash":"0x9a8c08120c2517fd2f5c96d377b401258121e3e44b9e5086ac236c6438bfac8e","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"0e1c6aac6663c379a52d9ccc7ba4757131020772d41447dfcf478cf9fb0c2bbf","Mgps":"151.160","gasUsed":"0x212496c6","time":3678536750,"test":"loopMul","fork":"London","d":2,"g":0,"v":0,"postHash":"0x221d03f92d6384bd7f947f156244f75262164d483dd1e11ee3f4c32e4030d9fe","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}```
```

New
```
{"output":"e2ef8b895ee03259ee57850b1a1ed4aa7689bec352659200e63440303f9d0b81","Mgps":"255.379","gasUsed":"0x96cd6dc2","time":9907033375,"test":"loopMul","fork":"Berlin","d":0,"g":0,"v":0,"postHash":"0x65134195701ec3aa354731a26d9533e946f11e11e60247cc96f57b305778fb12","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"ffffffffffffffffffffffffffaaffffffffffffffffbbffffffffffffff0009","Mgps":"336.091","gasUsed":"0x1778312c4","time":18745051292,"test":"loopMul","fork":"Berlin","d":1,"g":0,"v":0,"postHash":"0xa419240a3f3b2a7858212be7e9d6418d8cffd1673dca395582f191621555993d","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"0e1c6aac6663c379a52d9ccc7ba4757131020772d41447dfcf478cf9fb0c2bbf","Mgps":"191.400","gasUsed":"0x212496c6","time":2905156125,"test":"loopMul","fork":"Berlin","d":2,"g":0,"v":0,"postHash":"0x0ed14414e692f2ed97cef2b5a500863e2de358b42351d93b963ef4efb3d13d72","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"e2ef8b895ee03259ee57850b1a1ed4aa7689bec352659200e63440303f9d0b81","Mgps":"293.888","gasUsed":"0x96cd658e","time":8608883125,"test":"loopMul","fork":"Istanbul","d":0,"g":0,"v":0,"postHash":"0xee92d0c049225c53aa9b156986ad4e7cc9ec2524395c60876a17259fadc8aeae","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"ffffffffffffffffffffffffffaaffffffffffffffffbbffffffffffffff0009","Mgps":"334.783","gasUsed":"0x177830a90","time":18818286917,"test":"loopMul","fork":"Istanbul","d":1,"g":0,"v":0,"postHash":"0xe9bca424874ebc2457c9822f6c2f6d2b81ca9194b767c62f9b3c1e554d0a58f3","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"0e1c6aac6663c379a52d9ccc7ba4757131020772d41447dfcf478cf9fb0c2bbf","Mgps":"186.399","gasUsed":"0x21248e92","time":2983077875,"test":"loopMul","fork":"Istanbul","d":2,"g":0,"v":0,"postHash":"0xdf31131823f2d621cf2c2d980084681c7ac75a3991480c7d05a7031df34e21c5","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"e2ef8b895ee03259ee57850b1a1ed4aa7689bec352659200e63440303f9d0b81","Mgps":"293.697","gasUsed":"0x96cd6dc2","time":8614488584,"test":"loopMul","fork":"London","d":0,"g":0,"v":0,"postHash":"0x662588feffcd4391bc440510de21cbfa8f6c59ceafb2384ed706eccb55ec3a92","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"ffffffffffffffffffffffffffaaffffffffffffffffbbffffffffffffff0009","Mgps":"331.839","gasUsed":"0x1778312c4","time":18985231666,"test":"loopMul","fork":"London","d":1,"g":0,"v":0,"postHash":"0x9a8c08120c2517fd2f5c96d377b401258121e3e44b9e5086ac236c6438bfac8e","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
{"output":"0e1c6aac6663c379a52d9ccc7ba4757131020772d41447dfcf478cf9fb0c2bbf","Mgps":"186.066","gasUsed":"0x212496c6","time":2988433875,"test":"loopMul","fork":"London","d":2,"g":0,"v":0,"postHash":"0x221d03f92d6384bd7f947f156244f75262164d483dd1e11ee3f4c32e4030d9fe","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
```

**Comment by shemnon** [comment]
Work isn't done, I should finish unrolling the basic arithmetic that hasn't changed since Frontier.  Keeping the operations for complex operations won't have too bad an impact and preserves the customizability of the Besu EVM.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/JumpiOperation.java [hunk 1]
```diff
@@ -20,9 +20,6 @@
 import org.hyperledger.besu.evm.frame.MessageFrame;
 import org.hyperledger.besu.evm.gascalculator.GasCalculator;
 
-import java.util.Optional;
-import java.util.OptionalLong;
-
 import org.apache.tuweni.bytes.Bytes;
 
 public class JumpiOperation extends AbstractFixedCostOperation {
```
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/JumpiOperation.java:25]: Was there not a similar benefit to a static OperationResult(10,null) success response here?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/BaseFeeOperation.java [hunk 2]
```diff
@@ -34,8 +33,7 @@ public Operation.OperationResult executeFixedCostOperation(
       final MessageFrame frame, final EVM evm) {
     final Optional<Wei> maybeBaseFee = frame.getBlockValues().getBaseFee();
     if (maybeBaseFee.isEmpty()) {
-      return new Operation.OperationResult(
-          OptionalLong.of(gasCost), Optional.of(ExceptionalHaltReason.INVALID_OPERATION));
+      return new Operation.OperationResult(gasCost, ExceptionalHaltReason.INVALID_OPERATION);
     }
     frame.pushStackItem(maybeBaseFee.orElseThrow());
     return successResponse;
```
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/BaseFeeOperation.java:39]: same q about static OperationResult(2, null) here

File: evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperation.java [hunk 3]
```diff
@@ -46,4 +48,21 @@ public Operation.OperationResult executeFixedCostOperation(
 
     return successResponse;
   }
+
+  public static OperationResult staticOperation(final MessageFrame frame) {
```
Comment (garyschulte) [inline evm/src/main/java/org/hyperledger/besu/evm/operation/AddOperation.java:52]: Why are we not calling this from executeFixedCostOperation ?

**Review by garyschulte — APPROVED** [review]
non-blocking feedback and questions, but looks great.  It would be great to get this into the RC2 release today.

One other question - have you verified that  the pre/post stack change to tracing API results in no tracing output diffs?

File: evm/src/main/java/org/hyperledger/besu/evm/operation/JumpiOperation.java (hunk 1 above)
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/JumpiOperation.java:25]: no, there wasn't.  I was suprised.

File: evm/src/main/java/org/hyperledger/besu/evm/operation/BaseFeeOperation.java (hunk 2 above)
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/BaseFeeOperation.java:39]: Basefee gets overridden in downstream clients, so it cannot be optimized.

**Pull request merged**